addons/web/static/src/js/core/state_manager.js

```diff
@@ -0,0 +1,117 @@
+
+    /**
+     * Restores the state on each active subscriber assigned to the current type.
+     */
+    function restoreState() {
+        const { type, typeChanged } = currentContext;
+        if (!typeChanged) {
+            return;
+        }
+        for (const sub of subscriptions.values()) {
+            if (sub.alias in stateCache && sub.type === type) {
+                const state = parse(stateCache[sub.alias]);
+                if (sub.deferredState) {
+                    sub.pendingState = state;
+                } else {
+                    sub.import(state);
+                }
+            }
+        }
+        currentContext.typeChanged = false;
+    }
+
+    /**
+     * @typedef Context
+     * @param {action} number
+     * @param {type} string
+     */
+    /**
+     * Switches to a different context.
+     * If the action changes: the current subscriber map and cache are cleared.
+     * If the type changes: the current subscribers save their own state into
+     * the cache.
+     * @param {Context} context
+     */
+    function switchContext({ action, type }) {
+        // Checks outdated context
+        const actionChanged = action !== currentContext.action;
+        const typeChanged = type !== currentContext.type;
+        currentContext.typeChanged = typeChanged;
+        if (actionChanged) {
+            // Clears cache and subscriptions if exiting action
+            subscriptions.clear();
+            for (const key in stateCache) {
+                delete stateCache[key];
+            }
+        } else if (typeChanged) {
+            // Saves current subscriptions if changing type
+            for (const sub of subscriptions.values()) {
+                if (sub.type === currentContext.type) {
+                    stateCache[sub.alias] = stringify(sub.export());
+                }
+            }
+        }
+        // Assigns new value to the current state
+        Object.assign(currentContext, { action, type });
+    }
+
+    //-------------------------------------------------------------------------
+    // Hooks
+    //-------------------------------------------------------------------------
+
+    /**
+     * @typedef Handlers
+     * @property {Function} export
+     * @property {Function} import
+     */
+    /**
+     * Subscribe the current entity to a shared state identified by the given
+     * alias. After doing so the previous state will then be imported to the
+     * entity (this is ensured to be done right after complete constructor
+     * execution).
+     * @param {any} entity
+     * @param {string} alias
+     * @param {Handlers} handlers
+     */
+    function useSharedState(entity, alias, handlers) {
+        const { type } = currentContext;
+        const subscriber = Object.assign({ alias, type }, handlers);
+        const state = stateCache[alias] && parse(stateCache[alias]);
+        if (entity instanceof Component) {
+            subscriber.deferredState = true;
```

Comment:
I would maybe prefer isComponent or redo the test above in restoreState

Reply:
Done